sys/kern/vm_map.c

```diff
@@ -161,6 +161,46 @@ void vm_segment_destroy(vm_map_t *map, vm_segment_t *seg) {
   vm_segment_free(seg);
 }
 
+/* there are few cases:
+ * 1) we want to free the whole segment
+ * 2) we want to free the beginning of the segment and leave the end
+ * without changes
+ * 3) we want to free the end of the segment and leave the beginning
+ * without changes
+ * 4) we want to free the middle of the segment, so we want to split it to
+ * the new two segments with the hole between them */
+void vm_segment_destroy_range(vm_map_t *map, vm_segment_t *seg, vaddr_t start,
+                              vaddr_t end) {
+  assert(mtx_owned(&map->mtx));
+
+  /* case 1) */
+  if (seg->start == start && seg->end == end) {
+    vm_segment_destroy(map, seg);
+  } else if (seg->start == start) { /* case 2) */
+    assert(end < seg->end);
+    size_t length = end - seg->start;
+    vm_object_remove_range(seg->object, 0, length);
+    /* TODO there's no reference to pmap in page, so we have to do it here */
+    pmap_remove(map->pmap, seg->start, end);
+    seg->start = end;
+  } else if (seg->end == end) { /* case 3) */
+    assert(start > seg->start);
+    vm_segment_resize(map, seg, start);
+  } else { /* case 4) */
+    assert(start > seg->start && end < seg->end);
+    /* at the beginning: clone the vm_object */
+    vm_object_t *obj = vm_object_clone(seg->object);
+    /* nextly, resize the segment, to contain the beginning of it */
+    vm_segment_resize(map, seg, start);
+    /* now remove proper range from obj */
+    vm_object_remove_range(obj, 0, end - seg->start);
+    /* finally, create a new segment with obj */
+    vm_segment_t *new_seg =
+      vm_segment_alloc(obj, end, seg->end, seg->prot, seg->flags);
+    vm_map_insert_after(map, new_seg, seg);
+  }
```

Comment:
Could all this cases end in just one call to `vn_object_remove_range(obj, offset, length)`  where `offset` and `length` are calculated for all cases?

Could you explain why we resize segment in two cases but not in all (2,3,4) cases? It is a bit confusing to me.

Reply:
Okay, I finally did like you suggested: I did one call to vn_object_remove_range(obj, offset, length). And that changed everything so I think the second question is no longer relevant.